### PR TITLE
[ContactStructuralMechanicsApplication] Updating normal derivatives

### PR DESCRIPTION
**Pull request by loumalouomega**
WIP

Normal derivative corrected, added corresponding cpp tests

Now the NORMAL is a historical variable (needed to renormalize the normal after compute the derivative)

Some other corrections (parameters of tests updated to the historical NORMAL, etc...)

### CODE REVIEW

**Comment by loumalouomega** [comment]
@RiccardoRossi 80% of the files modified are parameters in json

File: kratos/utilities/mortar_utilities.h
Comment (RiccardoRossi) [inline kratos/utilities/mortar_utilities.h:906]: the original version was correct, why did u change that?

File: applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_frictional_mortar_contact_axisym_condition.h
Comment (RiccardoRossi) [inline applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_frictional_mortar_contact_axisym_condition.h:64]: why do you need it to be a template parameter? wouldn't it be better to pass it as a runtime option?

**Comment by loumalouomega** [comment]
PING

File: applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_frictional_mortar_contact_axisym_condition.cpp
Comment (RiccardoRossi) [inline applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_frictional_mortar_contact_axisym_condition.cpp]: why did u change from make_shared?

File: applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_mortar_contact_condition.h
Comment (RiccardoRossi) [inline applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_mortar_contact_condition.h:405]: why do you need to store a raw_pointer to the LeftHandSide matrix? what is it for?

File: applications/ContactStructuralMechanicsApplication/custom_conditions/mesh_tying_mortar_condition.cpp
Comment (RiccardoRossi) [inline applications/ContactStructuralMechanicsApplication/custom_conditions/mesh_tying_mortar_condition.cpp:1271]: you have twice template<>

File: applications/ContactStructuralMechanicsApplication/custom_processes/alm_fast_init_process.cpp
Comment (RiccardoRossi) [inline applications/ContactStructuralMechanicsApplication/custom_processes/alm_fast_init_process.cpp:39]: take care, ZeroVector is an expression, not a vector. i would use the ZeroVector inside instead of allocating an object

File: applications/ContactStructuralMechanicsApplication/custom_strategies/custom_convergencecriterias/alm_frictional_mortar_criteria.h
Comment (RiccardoRossi) [inline applications/ContactStructuralMechanicsApplication/custom_strategies/custom_convergencecriterias/alm_frictional_mortar_criteria.h:221]: i am wondering if using openmp reduction for those variables would work here. I am not sure, but if it does it would probably be faster

File: applications/ContactStructuralMechanicsApplication/custom_utilities/contact_utilities.h
Comment (RiccardoRossi) [inline applications/ContactStructuralMechanicsApplication/custom_utilities/contact_utilities.h:156]: noalias

File: applications/ContactStructuralMechanicsApplication/custom_utilities/contact_utilities.h
Comment (RiccardoRossi) [inline applications/ContactStructuralMechanicsApplication/custom_utilities/contact_utilities.h:223]: use ZeroVector instead. It is an expression!!

File: applications/ContactStructuralMechanicsApplication/custom_utilities/derivatives_utilities.h
Comment (RiccardoRossi) [inline applications/ContactStructuralMechanicsApplication/custom_utilities/derivatives_utilities.h:534]: NO_DERIVATIVES_COMPUTATION would be better

File: kratos/includes/mortar_classes.h
Comment (RiccardoRossi) [inline kratos/includes/mortar_classes.h:504]: what are u doing here?

File: applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_frictional_mortar_contact_axisym_condition.cpp
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_frictional_mortar_contact_axisym_condition.cpp]: to remove the boost:: dependency

File: applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_frictional_mortar_contact_axisym_condition.h
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_frictional_mortar_contact_axisym_condition.h:64]: Yes, because the dependency is defined in the AD, the number of terms with the definition is reduced considerably

File: applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_mortar_contact_condition.h
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_mortar_contact_condition.h:405]: Just to keep the structure what it was done previously

File: applications/ContactStructuralMechanicsApplication/custom_conditions/mesh_tying_mortar_condition.cpp
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/custom_conditions/mesh_tying_mortar_condition.cpp:1271]: Because there are two template definition, the method and the class

File: applications/ContactStructuralMechanicsApplication/custom_processes/alm_fast_init_process.cpp
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/custom_processes/alm_fast_init_process.cpp:39]: OK

File: applications/ContactStructuralMechanicsApplication/custom_strategies/custom_convergencecriterias/alm_frictional_mortar_criteria.h
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/custom_strategies/custom_convergencecriterias/alm_frictional_mortar_criteria.h:221]: I think I already checked and didn't work

File: applications/ContactStructuralMechanicsApplication/custom_utilities/contact_utilities.h
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/custom_utilities/contact_utilities.h:156]: Yep

File: applications/ContactStructuralMechanicsApplication/custom_utilities/contact_utilities.h
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/custom_utilities/contact_utilities.h:223]: Ok

File: applications/ContactStructuralMechanicsApplication/custom_utilities/derivatives_utilities.h
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/custom_utilities/derivatives_utilities.h:534]: Ok

File: kratos/includes/mortar_classes.h
Comment (loumalouomega) [reply kratos/includes/mortar_classes.h:504]: Defining and zero size matrix in the cases is not use, to reduce the amount of memory used

File: kratos/utilities/mortar_utilities.h
Comment (loumalouomega) [reply kratos/utilities/mortar_utilities.h:906]: I assumed that using the initialization of the class was faster

**Comment by loumalouomega** [comment]
@RiccardoRossi Can I correct your considerations in the future PR with the assembling modification?, I changed so  many things in this new branch that I have conflicts if I try to do the modifications suggested and cherrypick

**Comment by loumalouomega** [comment]
PING I did the modifications in the new branch, but there are so many things changes that it is impossible to cherry-pick. Can you approve please?

**Comment by loumalouomega** [comment]
PING

File: kratos/utilities/mortar_utilities.h
Comment (RiccardoRossi) [reply kratos/utilities/mortar_utilities.h:906]: you are not necessarily right on this. anyway, the new code is also  correct

File: applications/ContactStructuralMechanicsApplication/symbolic_generation/ALM_frictional_mortar_condition/ALM_frictional_mortar_contact_condition.cpp
Comment (RiccardoRossi) [inline applications/ContactStructuralMechanicsApplication/symbolic_generation/ALM_frictional_mortar_condition/ALM_frictional_mortar_contact_condition.cpp]: can i ask why you switched back to the Pointer construction? the make_shared should be better

File: kratos/utilities/exact_mortar_segmentation_utility.cpp
Comment (RiccardoRossi) [inline kratos/utilities/exact_mortar_segmentation_utility.cpp]: you can not change this as such... it is in the core and it may be used by other people.

Is there a test in the core to check this btw?

File: kratos/utilities/mortar_utilities.h
Comment (RiccardoRossi) [inline kratos/utilities/mortar_utilities.h:184]: this is also in the core. You can not change it as you wish...

**Review by RiccardoRossi — COMMENTED** [review]
you changed somethign in the core. Did u ensure that this is not used by other people? also, do you really need the NORMAL to be a SolutionStepVariable instead of a nonhisotrical variable as it was before?

File: kratos/utilities/mortar_utilities.h
Comment (loumalouomega) [reply kratos/utilities/mortar_utilities.h:184]: Right now I am the only one using it. I put it in here for the future developments of the MappingApplication, not yet used

File: kratos/utilities/exact_mortar_segmentation_utility.cpp
Comment (loumalouomega) [reply kratos/utilities/exact_mortar_segmentation_utility.cpp]: There are a very simple python test in the core. The cpp tests and the most advances tests are in my application because depend of the search (that is in my application)

File: kratos/utilities/mortar_utilities.h
Comment (loumalouomega) [reply kratos/utilities/mortar_utilities.h:906]: In any case this is modified in the new branch

File: applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_frictional_mortar_contact_axisym_condition.cpp
Comment (RiccardoRossi) [reply applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_frictional_mortar_contact_axisym_condition.cpp]: make_shared has a higher performance at creation and destruction time.

you are creating/destructing this condition many times...so here it could make a difference

File: kratos/includes/mortar_classes.h
Comment (RiccardoRossi) [inline kratos/includes/mortar_classes.h:492]: once this is in the core, it should be tested in the core. 
Also you CAN NOT change this as you wish starting from the moment in which it is there.

i am going to accept this for now, but next time you want to change this you must do a PR on its own, just changing this file and the relevant tests.

File: kratos/utilities/exact_mortar_segmentation_utility.cpp
Comment (RiccardoRossi) [reply kratos/utilities/exact_mortar_segmentation_utility.cpp]: then you should make a cpp test which does not depend on that and put it in the core... otherwise it does not make sense to have it there

File: kratos/utilities/mortar_utilities.h
Comment (RiccardoRossi) [reply kratos/utilities/mortar_utilities.h:184]: same comment. Then be sure that the tests are there. i will accept for this time, but next time you want to change it should be in a separate PR

**Review by RiccardoRossi — DISMISSED** [review]
i am approving even if i am not completely happy with this.

the two important comments are
1 - the make_shared is important for performance
2 - for the future you cannot change the mortar classes one sidedly

File: kratos/includes/mortar_classes.h
Comment (loumalouomega) [reply kratos/includes/mortar_classes.h:492]: Ok

**Comment by loumalouomega** [comment]
We should make a macro for the creation, or it will be deprecated as soon as we change the python library

**Comment by loumalouomega** [comment]
I will correct the conflict before merge

**Comment by loumalouomega** [comment]
Here is @RiccardoRossi I solve the conflict (due to Coordinate()). I updated the application to the new standard and I restores the make_shared. Besides I update the python test for mortar segmentation

**Review by RiccardoRossi — APPROVED** [review]

**Pull request merged**
